### ISSUE TITLE
CCD problem

### ISSUE THREAD

**Issue by GoogleCodeExporter** [erincatto/box2d#220]
```
See the attached testbed file.

When a large heavy bullet plows through a bunch of smaller, lighter objects at 
high speeds, it appears to stutter as it goes through.  The CCD system seems to 
be giving up after only solving a small portion of the step (it's not entirely 
clear, but single-stepping makes it look like it might be solving only a single 
impact per step).

Obviously there needs to be a limit to the amount of computation CCD will do 
per step, but this doesn't appear to be tunable in the current implementation 
(b2_maxTOIContacts doesn't seem to affect this problem), so if it's possible 
and doesn't require a lot of algorithmic change, a good fix would be to make 
whatever is limiting this tunable in b2Settings.h.
```

Original issue reported on code.google.com by `ewjor...@gmail.com` on 10 Jul 2011 at 7:16

Attachments:
- [CCDProblem.h](https://storage.googleapis.com/google-code-attachments/box2d/issue-220/comment-0/CCDProblem.h)

**Comment by louis-langholtz** [comment]
@erincatto If you're interested, I did some experimentation with algorithmic changes and using different data structures for this. The algorithmic change was just using the cached element counts I had available for both the contact associated bodies and picking the body to search over that had the least contacts. Data structure wise, I tried `std::list`, `std::set`, `std::unordered_set`, and `std::vector` data structures.

What I found surprised me.

Relatively speaking, I was seeing max per-step times with 400 smaller elements of around 0.6 seconds (usually at step 15 in my trials using the code in my fork). And that was using a `std::set` member within the world instance. `std::unordered_set` lowered that to around 0.5 seconds which turned out to be about the same times as I was seeing when `std::unordered_set` was instead based on the smaller per-body contacts list. Also about the same time as just using `std::list` of contacts on the bodies.

What turned out consistently fastest for me, was using `std::vector` of contacts on the bodies. That had an average max time at step 15 of about 0.2 seconds. I didn't just push the contacts though when using `std::list` or `std::vector`. I also added the contact identifying information to the list and the vector so that the code wouldn't have to re-lookup through possibly cold memory the identifying information. I didn't try to confirm that the vector would be slower if its elements didn't include this key info. I then got worried that maybe the vector would suffer significantly from removing contact elements so I looked at the sum of step times out to step 20. The vector still outperformed the other methods. Vector was about twice as fast as the next best method.
